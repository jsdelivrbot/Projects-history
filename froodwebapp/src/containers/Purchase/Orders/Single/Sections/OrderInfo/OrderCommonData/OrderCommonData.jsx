import React from 'react';
import PropTypes from 'prop-types';
import { Row, Col } from 'react-flexbox-grid';
import {
  FormDatePicker,
  FormSelect,
  ColouredStatus
} from 'components';
import { Field } from 'redux-form';
import { getMenuItems } from 'utils';

const OrderCommonData = ({
  // status
  status,
  statusColor,

  // select data
  shippingMethods,
  payterms,
  vendorPaymentTermExist,
  contactUsers,

  // form data

  // state
  isNewOrder,
  disabled
}) => (
  <Col xs md={ 6 } lg={ 5 }>
    <Row middle="xs">
      <Col xs sm md lg>
        <ColouredStatus
          color={ statusColor }
          status={ status }
        />
      </Col>
    </Row>
    { !isNewOrder &&
      <Row middle="xs" end="xs">
        <Col>
          <strong>Purchase Order Date</strong>
        </Col>
        <Col xs md={ 8 } lg={ 4 }>
          <Field
            name="date"
            component={ FormDatePicker }
            disabled
          />
        </Col>
      </Row>
    }
    <Row middle="xs" end="xs">
      <Col>
        <strong>
          Required By
        </strong>
      </Col>
      <Col xs md={ 8 } lg={ 4 }>
        <Field
          name="deliveryDate"
          component={ FormDatePicker }
          disabled={ disabled }
        />
      </Col>
    </Row>
    <Row middle="xs" end="xs">
      <Col>
        <strong>Shipping Method</strong>
      </Col>
      <Col xs md={ 8 } lg={ 6 }>
        <Field
          name="deliveryMethod.id"
          component={ FormSelect }
          menuItems={ getMenuItems(shippingMethods) }
          disabled={ disabled }
        />
      </Col>
    </Row>
    <Row middle="xs" end="xs">
      <Col>
        <strong>Payment Term</strong>
      </Col>
      <Col xs md={ 8 } lg={ 6 }>
        <Field
          name="paymentTerms.id"
          component={ FormSelect }
          menuItems={ getMenuItems(payterms) }
          disabled={ vendorPaymentTermExist }
        />
      </Col>
    </Row>
    <Row middle="xs" end="xs">
      <Col>
        <strong>Contact User</strong>
      </Col>
      <Col xs md={ 8 } lg={ 6 }>
        <Field
          name="contactUser.id"
          component={ FormSelect }
          menuItems={ getMenuItems(contactUsers) }
          disabled={ disabled }
        />
      </Col>
    </Row>
  </Col>
);

OrderCommonData.propTypes = {
  // common data
  shippingMethods: PropTypes.array,
  payterms: PropTypes.array,
  vendorPaymentTermExist: PropTypes.bool.isRequired,
  contactUsers: PropTypes.array,
  // props
  status: PropTypes.string,
  isNewOrder: PropTypes.bool.isRequired,
  disabled: PropTypes.bool.isRequired,
  statusColor: PropTypes.string
};

export default OrderCommonData;
