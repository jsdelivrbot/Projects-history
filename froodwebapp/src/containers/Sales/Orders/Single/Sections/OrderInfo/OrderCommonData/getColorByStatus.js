const colorsByStatus = [{
  color: '#34A5DA',
  status: 'Draft',
}, {
  color: '#4C8945',
  status: 'Confirmed'
}, {
  color: '#838787',
  status: 'Backorder'
}, {
  color: '#F96928',
  status: 'Allocated'
}, {
  color: '#FF9300',
  status: 'PartialAllocated'
}, {
  color: '#0096FF',
  status: 'PickProgress'
}, {
  color: '#0433FF',
  status: 'Picked'
}, {
  color: '#9437FF',
  status: 'PackProgress'
}, {
  color: '#FF2F92',
  status: 'Packed'
}, {
  color: '#FF7E79',
  status: 'Shipped'
}, {
  color: '#000000',
  status: 'Delivered'
}, {
  color: '#C0C0C0',
  status: 'Invoiced'
}, {
  color: '#929292',
  status: 'PaymentProgress'
}, {
  color: '#5E5E5E',
  status: 'Paid'
}, {
  color: '#FF2600',
  status: 'FailedPayment'
}];

export default (status) => {
  const colorItem = colorsByStatus.find(item => item.status.toLowerCase() === status.toLowerCase());

  return colorItem ? colorItem.color : 'black';
};
