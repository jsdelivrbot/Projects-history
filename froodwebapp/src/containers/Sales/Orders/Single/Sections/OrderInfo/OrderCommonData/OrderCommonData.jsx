import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Row, Col } from 'react-flexbox-grid';
import moment from 'moment';
import { connect } from 'react-redux';
import {
  FormDatePicker,
  FormSelect,
  FormInput,
  ColouredStatus
} from 'components';
import { Field, formValueSelector } from 'redux-form';
import { deliveryInfoParallelRequest } from 'redux-base/actions';
import { getMenuItems } from 'utils';
import getColorByStatus from './getColorByStatus';

const selector = formValueSelector('newOrderForm');

const mapStateToProps = state => ({
  payterms: state.commonData.payterms,
  deliveryMethods: state.commonData.deliveryMethods,
  deliveryMethodInfo: state.commonData.deliveryMethodInfo,
  deliverySlotsInfo: state.commonData.deliverySlotsInfo,
  status: state.order.initialValues.status,
  channelId: state.order.initialValues.channel.id,
  defaultCustomerAddress: state.order.defaultCustomerAddress,
  customerName: selector(state, 'customer.name'),
  deliveryDate: selector(state, 'deliveryDate'),
  deliveryMethodId: selector(state, 'deliveryMethodId')
});

const mapDispatchToProps = { deliveryInfoParallelRequest };

export class OrderCommonData extends Component {
  componentWillReceiveProps(nextProps) {
    const {
      disabled,
      deliveryMethodId,
      deliveryDate,
      defaultCustomerAddress,
      customerName
    } = nextProps;


    if (!disabled) {
      // load slots and delivery days when delivery method is selected
      if (deliveryMethodId
        && deliveryMethodId !== this.props.deliveryMethodId) {
        this.props.deliveryInfoParallelRequest({
          params: [{
            id: deliveryMethodId
          }, {
            id: deliveryMethodId
          }]
        });
      }

      // Reset delivery slot on delivery date change
      if (this.props.deliveryDate
        && deliveryDate !== this.props.deliveryDate) {
        this.props.onFieldChange('deliverySlotId', null);
      }

      // initialize shipping and billing address when default customer address loaded
      if (defaultCustomerAddress
       && defaultCustomerAddress.id !== this.props.defaultCustomerAddress.id) {
        this.props.onFieldChange('shippingAddress', {
          id: defaultCustomerAddress.id,
          name: customerName,
          address: defaultCustomerAddress
        });
        this.props.onFieldChange('billingAddress', {
          id: defaultCustomerAddress.id,
          name: customerName,
          address: defaultCustomerAddress
        });
      }
    }
  }

  render() {
    const {
      // form
      status,
      channelId,
      deliveryMethodId,
      deliveryDate,
      // static data
      deliveryMethods,
      deliveryMethodInfo,
      deliverySlotsInfo,
      payterms,
      // state
      disabled
    } = this.props;

    let deliveryMethodsMenuItems = [];
    let slotsMenuItems = [];
    let paymentTermsMenuItems = [];
    let firstActiveDeliveryDate = null;
    let lastActiveDeliveryDate = null;


    if (disabled) {
      // load all select data, even disabled or not active so select values can be populated
      deliveryMethodsMenuItems = getMenuItems(deliveryMethods);
      paymentTermsMenuItems = getMenuItems(payterms);
      slotsMenuItems = getMenuItems(deliveryMethods.find(dm => dm.id === deliveryMethodId).slots);
    } else {
      deliveryMethodsMenuItems = getMenuItems(deliveryMethods.filter(dm => dm.isActive));

      // get slots by delivery date
      if (deliveryMethodId && deliveryDate) {
        const deliveryInfo = deliverySlotsInfo.find(dm => dm.id === moment(deliveryDate, 'DD-MMMM-YYYY').isoWeekday());
        if (deliveryInfo && deliveryInfo.slots) {
          slotsMenuItems = deliveryInfo.slots.map(slot => ({
            key: slot.id,
            value: `${slot.from} - ${slot.to}`
          }));
        }
      }

      // get delivery days by delivery method
      if (deliveryMethodId && deliveryMethodInfo.calLeadDays) {
        const { calLeadDays, calVisibleDays } = deliveryMethodInfo;

        firstActiveDeliveryDate = moment().add(calLeadDays, 'd').startOf('day');
        lastActiveDeliveryDate = moment().add(calLeadDays + calVisibleDays, 'd').startOf('day');
      }

      // get payment terms by channel
      paymentTermsMenuItems = getMenuItems(payterms.filter(pt => pt.channels && pt.channels.includes(channelId)));
    }

    const statusColor = getColorByStatus(status);

    return (
      <div>
        <Row middle="xs">
          <Col xs sm md lg>
            <ColouredStatus
              color={ statusColor }
              status={ status }
            />
          </Col>
        </Row>
        <Row middle="xs" end="xs">
          <Col xs>
            <strong>
              Order Channel
            </strong>
          </Col>
          <Col xs md={ 8 } lg={ 4 }>
            <Field
              name="channel.name"
              disabled
              component={ FormInput }
            />
          </Col>
        </Row>
        <Row middle="xs" end="xs">
          <Col xs>
            <strong>
              Channel Ref
            </strong>
          </Col>
          <Col xs md={ 8 } lg={ 4 }>
            <Field
              name="channelRefNo"
              disabled
              component={ FormInput }
            />
          </Col>
        </Row>
        <Row middle="xs" end="xs">
          <Col xs>
            <strong>Order Date</strong>
          </Col>
          <Col xs md={ 8 } lg={ 4 }>
            <Field
              name="orderDate"
              disabled={ disabled }
              activeFromToday
              component={ FormDatePicker }
            />
          </Col>
        </Row>
        <Row middle="xs" end="xs">
          <Col>
            <strong>Payment Terms</strong>
          </Col>
          <Col xs md={ 8 } lg={ 4 }>
            <Field
              name="paymentTerms.id"
              placeholder="Payment Term"
              component={ FormSelect }
              menuItems={ paymentTermsMenuItems }
              disabled={ disabled }
            />
          </Col>
        </Row>
        <Row middle="xs" end="xs">
          <Col>
            <strong>Delivery Method</strong>
          </Col>
          <Col xs md={ 8 } lg={ 4 }>
            <Field
              name="deliveryMethodId"
              placeholder="Delivery Method"
              component={ FormSelect }
              menuItems={ deliveryMethodsMenuItems }
              disabled={ disabled }
            />
          </Col>
        </Row>
        <Row middle="xs" end="xs">
          <Col>
            <strong>
              Expected Delivery Date
            </strong>
          </Col>
          <Col xs md={ 8 } lg={ 4 }>
            <Field
              name="deliveryDate"
              firstActiveDate={ firstActiveDeliveryDate }
              lastActiveDate={ lastActiveDeliveryDate }
              component={ FormDatePicker }
              disabled={ disabled || !deliveryMethodId }
            />
          </Col>
        </Row>
        <Row middle="xs" end="xs">
          <Col>
            <strong>Delivery Slot</strong>
          </Col>
          <Col xs md={ 8 } lg={ 4 }>
            <Field
              name="deliverySlotId"
              placeholder="Delivery Slot"
              component={ FormSelect }
              menuItems={ slotsMenuItems }
              disabled={ disabled || !deliveryDate }
            />
          </Col>
        </Row>
      </div>
    );
  }
}

OrderCommonData.propTypes = {
  // static data
  payterms: PropTypes.array,
  deliveryMethods: PropTypes.array,
  deliveryMethodInfo: PropTypes.object,
  deliverySlotsInfo: PropTypes.array,

  // props
  defaultCustomerAddress: PropTypes.object,

  // form data
  channelId: PropTypes.number.isRequired,
  status: PropTypes.string.isRequired,
  deliveryMethodId: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  deliveryDate: PropTypes.string,
  customerName: PropTypes.string,
  // redux-form
  onFieldChange: PropTypes.func.isRequired,
  // redux-base
  deliveryInfoParallelRequest: PropTypes.func.isRequired,
  // state data
  disabled: PropTypes.bool.isRequired,
};

export default connect(mapStateToProps, mapDispatchToProps)(OrderCommonData);
