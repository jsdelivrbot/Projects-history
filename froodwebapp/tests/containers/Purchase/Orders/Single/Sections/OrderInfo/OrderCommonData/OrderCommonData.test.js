import OrderCommonData from 'containers/Purchase/Orders/Single/Sections/OrderInfo/OrderCommonData/OrderCommonData';
import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

const { describe, expect, it } = global;

describe('OrderCommonData', () => {
  const props = {
    shippingMethods: [],
    payterms: [],
    contactUsers: [],
    status: 'status',
    isNewOrder: false,
    readonly: false,
    canBeEditedUntilPick: false,
    statusColor: 'status',
    disabled: false
  };

  it('renders the OrderCommonData component', () => {
    const component = shallow(<OrderCommonData { ...props } />);

    expect(component.length).toEqual(1);
  });

  it('capturing Snapshot of OrderCommonData', () => {
    const component = shallow(<OrderCommonData { ...props } />);
    const view = toJson(component);
    const actual = expect(view);

    actual.toMatchSnapshot();
  });
});
