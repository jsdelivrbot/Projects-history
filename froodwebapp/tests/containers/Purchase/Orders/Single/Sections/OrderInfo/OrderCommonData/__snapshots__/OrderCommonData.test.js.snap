// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderCommonData capturing Snapshot of OrderCommonData 1`] = `
<Col
  lg={5}
  md={6}
  xs={true}
>
  <Row
    end="xs"
    middle="xs"
  >
    <Col
      lg={4}
      md={4}
      sm={6}
      xs={true}
    >
      <div
        style={
          Object {
            "backgroundColor": "status",
            "borderRadius": "1rem",
            "color": "white",
            "padding": "0.4rem 2rem",
            "textAlign": "center",
          }
        }
      >
        status
      </div>
    </Col>
  </Row>
  <Row
    end="xs"
    middle="xs"
  >
    <Col>
      <strong>
        Purchase Order Date
      </strong>
    </Col>
    <Col
      lg={4}
      md={8}
      xs={true}
    >
      <Field
        component={[Function]}
        disabled={true}
        name="date"
      />
    </Col>
  </Row>
  <Row
    end="xs"
    middle="xs"
  >
    <Col>
      <strong>
        Required By
      </strong>
    </Col>
    <Col
      lg={4}
      md={8}
      xs={true}
    >
      <Field
        component={[Function]}
        disabled={false}
        name="deliveryDate"
      />
    </Col>
  </Row>
  <Row
    end="xs"
    middle="xs"
  >
    <Col>
      <strong>
        Shipping Method
      </strong>
    </Col>
    <Col
      lg={6}
      md={8}
      xs={true}
    >
      <Field
        component={[Function]}
        disabled={false}
        menuItems={Array []}
        name="deliveryMethod.id"
      />
    </Col>
  </Row>
  <Row
    end="xs"
    middle="xs"
  >
    <Col>
      <strong>
        Payment Term
      </strong>
    </Col>
    <Col
      lg={6}
      md={8}
      xs={true}
    >
      <Field
        component={[Function]}
        disabled={true}
        menuItems={Array []}
        name="paymentTerms.id"
      />
    </Col>
  </Row>
  <Row
    end="xs"
    middle="xs"
  >
    <Col>
      <strong>
        Contact User
      </strong>
    </Col>
    <Col
      lg={6}
      md={8}
      xs={true}
    >
      <Field
        component={[Function]}
        disabled={false}
        menuItems={Array []}
        name="contactUser.id"
      />
    </Col>
  </Row>
</Col>
`;
